src/ReactiveUI/Suspension/SuspensionHostExtensions.cs

```diff
@@ -65,32 +81,65 @@ public static IDisposable SetupDefaultSuspendResume(this ISuspensionHost item, I
         }
 
         var ret = new CompositeDisposable();
-        driver ??= Locator.Current.GetService<ISuspensionDriver>();
+        suspensionDriver ??= Locator.Current.GetService<ISuspensionDriver>();
 
-        if (driver is null)
+        if (suspensionDriver is null)
         {
             item.Log().Error("Could not find a valid driver and therefore cannot setup Suspend/Resume.");
             return Disposable.Empty;
         }
 
         ret.Add(item.ShouldInvalidateState
-                    .SelectMany(_ => driver.InvalidateState())
+                    .SelectMany(_ => suspensionDriver.InvalidateState())
                     .LoggedCatch(item, Observables.Unit, "Tried to invalidate app state")
                     .Subscribe(_ => item.Log().Info("Invalidated app state")));
 
         ret.Add(item.ShouldPersistState
-                    .SelectMany(x => driver.SaveState(item.AppState!).Finally(x.Dispose))
+                    .SelectMany(x => suspensionDriver.SaveState(item.AppState!).Finally(x.Dispose))
                     .LoggedCatch(item, Observables.Unit, "Tried to persist app state")
                     .Subscribe(_ => item.Log().Info("Persisted application state")));
 
         ret.Add(item.IsResuming.Merge(item.IsLaunchingNew)
-                    .SelectMany(_ => driver.LoadState())
-                    .LoggedCatch(
-                                 item,
-                                 Observable.Defer(() => Observable.Return(item.CreateNewAppState?.Invoke())),
-                                 "Failed to restore app state from storage, creating from scratch")
-                    .Subscribe(x => item.AppState = x ?? item.CreateNewAppState?.Invoke()));
+                    .SelectMany(_ => item.EnsureLoadAppState(suspensionDriver))
+                    .Subscribe());
 
         return ret;
     }
-}
\ No newline at end of file
+
+    /// <summary>
+    /// Ensures one time app state load from storage.
+    /// </summary>
+    /// <param name="item">The suspension host.</param>
+    /// <param name="driver">The suspension driver.</param>
+    /// <returns>A completed observable.</returns>
+    private static IObservable<Unit> EnsureLoadAppState(this ISuspensionHost item, ISuspensionDriver? driver = null)
```

Comment:
You can possibly use the `Interlocked.Exchange(ref initFunction, null)?.Invoke()` then store a Action or Func with the init logic. This is thread safe as well but avoids the lock

Reply:
Oh wow, it took me a while, I have never used `Interlocked`, but I understand and this is clever. Commit will be coming right up

Reply:
Looks like you got the idea correctly.